### PR TITLE
Change empty message to an element with attributes

### PR DESCRIPTION
**Pull request by samuelli**
This adds a debouncer to the base element. When observed attributes are changed, `renderCallback()` is called, where `observedAttributes()` is defined by the element implementation.

@cdata PTAL at `base-element` changes.

### CODE REVIEW

File: src/client/public/scripts/components/base-element.ts [hunk 1]
```diff
@@ -1,17 +1,44 @@
...
   }
 
   connectedCallback() {
     this.renderCallback();
   }
 
   renderCallback() {
-    render(this.render(), this);
+    this._debouncer(() => {
+      render(this.render(), this);
```
Comment (gauntface) [inline src/client/public/scripts/components/base-element.ts]: could you rename lit's render to `litRender`.

File: src/client/public/scripts/components/base-element.ts [hunk 2]
```diff
@@ -1,17 +1,44 @@
...
+type Debouncer = (task: Function) => Promise<void>;
+
+function createDebouncer(): Debouncer {
+  let queued = false;
+  return async (task: Function) => {
+    if (queued) {
+      return;
+    }
+    queued = true;
+    await Promise.resolve();
```
Comment (gauntface) [inline src/client/public/scripts/components/base-element.ts]: Can you add some comments on why you are await a promise.resolve - I'm assuming to allow microtasks to finish before rendering.

Is there a reason this is preferable over rAF?

File: src/client/public/scripts/components/base-element.ts [hunk 3]
```diff
@@ -1,17 +1,44 @@
...
+    queued = false;
+    task();
+  };
+}
+
 export abstract class BaseElement extends HTMLElement {
+  private _debouncer: Debouncer;
```
Comment (cdata) [inline src/client/public/scripts/components/base-element.ts]: With TypeScript you could even initialize this inline and avoid writing out the constructor.

File: src/client/public/scripts/components/base-element.ts (hunk 2 above)
Comment (cdata) [reply src/client/public/scripts/components/base-element.ts]: In the microtask case you are avoiding rerender after every render call in a single turn. The raf case also works as long as you are okay with waiting the extra time to render. Microtask is just the minimum viable timing to use for this model IMO.

File: src/client/public/scripts/components/base-element.ts (hunk 1 above)
Comment (samuelli) [reply src/client/public/scripts/components/base-element.ts]: Yes!

File: src/client/public/scripts/components/base-element.ts (hunk 2 above)
Comment (samuelli) [reply src/client/public/scripts/components/base-element.ts]: What he said.

File: src/client/public/scripts/components/base-element.ts (hunk 3 above)
Comment (samuelli) [reply src/client/public/scripts/components/base-element.ts]: ++, done.

**Pull request merged**